### PR TITLE
Improve XLCell.SetValue(object)

### PR DESCRIPTION
**Pull request by igitur**
Improve XLCell.SetValue(object) by removing unnecessary TryParse functions

### CODE REVIEW

**Comment by Pankraty** [comment]
Please wait before #679 is merged. I rebased it ~6 times already )

**Comment by igitur** [comment]
So you should be able to rebase in your sleep by now!

>

**Comment by Pankraty** [comment]
:-P 
I do actually, but with rebasing, I should resolve merge conflicts too ) That's not fair!

**Comment by vbjay** [comment]
https://hackernoon.com/fix-conflicts-only-once-with-git-rerere-7d116b2cec67

**Comment by Pankraty** [comment]
Thanks, @vbjay! In this particular case, the problem was that the same code has been modified for a few times during the PR life, but in some other PRs this trick will be very useful.

**Comment by igitur** [comment]
This PR is ready for review. `XLCell.SetValue` has been refactored substantially, so the diffs won't make sense. I do believe the method is a lot more readable now.

File: ClosedXML_Tests/Excel/Cells/XLCellTests.cs [hunk 1]
```diff
@@ -369,6 +369,36 @@ public void ValueSetDateWithShortUserDateFormat()
             Assert.AreEqual(expected, actual);
         }
 
+        [Test]
+        public void SetStringCellValues()
```
Comment (Pankraty) [inline ClosedXML_Tests/Excel/Cells/XLCellTests.cs]: I would suggest adding a few more test cases:

* Specify an empty string
* Specify a string longer than 32767 characters
* Specify null
* Specify number as a string
* Specify DateTime less than `BaseDate`
* Specify DateTime, Decimal, Integer, TimeSpan, Boolean without converting to strings
* Specify `FormulaA1` and then set cell value and ensure that both `FormulaA1` and `FormulaR1C1` are empty
* Specify a string having line break included and ensure that `WrapText` is set to true
* Specify a string containing a single quote and number (e.g. `'10`) and make sure the `DataType` is Text.

Perhaps some of those cases are already covered with tests but they can help not to get into regression in future.

**Comment by igitur** [comment]
Ready for another review. Made more improvements.

**Comment by igitur** [comment]
Breaking change: setting a cell value using `.SetValue()` to a date that's lower than the base date will throw an `ArgumentException`.

File: ClosedXML/Excel/Cells/XLCell.cs [hunk 2]
```diff
@@ -218,57 +218,125 @@ internal IXLCell SetValue<T>(T value, bool setTableHeader)
...
+            {
+                parsedValue = b ? "1" : "0";
+                _dataType = XLDataType.Boolean;
+                parsed = true;
+            }
+            else if (value.IsNumber())
+            {
+                if ((value is double || value is float)
+                    && (Double.IsNaN((Double)Convert.ChangeType(value, typeof(Double)))
+                        || Double.IsInfinity((Double)Convert.ChangeType(value, typeof(Double)))))
```
Comment (Pankraty) [inline ClosedXML/Excel/Cells/XLCell.cs]:
```
if ((value is double d && (double.IsNaN(d) || double.IsInfinity(d)) ||
    (value is float f && (float.IsNaN(f) || float.IsInfinity(f))
```
looks cleaner and does not require `(Double)Convert.ChangeType(value, typeof(Double))` to be executed twice.

File: ClosedXML/Excel/Cells/XLCell.cs [hunk 3]
```diff
@@ -218,57 +218,125 @@ internal IXLCell SetValue<T>(T value, bool setTableHeader)
...
+                throw new ArgumentException($"Unable to set cell value to {value.ToInvariantString()}", nameof(value));
+
+            SetInternalCellValueString(parsedValue, validate: true, parseToCachedValue: false);
+
+            CachedValue = null;
+
+            return this;
+        }
+
+        private Tuple<string, bool> SetKnownTypedValue<T>(T value, XLStyleValue style)
```
[the remark this answers is not in the document]
Comment (igitur) [reply ClosedXML/Excel/Cells/XLCell.cs]: Tried that, but requires new libraries to be referenced (ValueTuple) and couldn't get it working for netstandard, so left it deliberately until we implement ValueTuple in many places.

File: ClosedXML/Excel/Cells/XLCell.cs (hunk 2 above)
Comment (igitur) [reply ClosedXML/Excel/Cells/XLCell.cs]: True. And I must add a test for this.

File: ClosedXML/Excel/Cells/XLCell.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Pankraty) [reply ClosedXML/Excel/Cells/XLCell.cs]: Got it.

File: ClosedXML/Excel/Cells/XLCell.cs [hunk 4]
```diff
@@ -434,12 +502,18 @@ public Object Evaluate(Boolean force = false)
...
-            SetInternalCellValueString(cellValue, this.HasFormula);
+            SetInternalCellValueString(cellValue, validate: false, parseToCachedValue: this.HasFormula);
         }
 
-        private void SetInternalCellValueString(String cellValue, Boolean parseToCachedValue)
+        private void SetInternalCellValueString(String cellValue, Boolean validate, Boolean parseToCachedValue)
         {
+            if (validate)
+            {
+                if (cellValue.Length > 32767) throw new ArgumentException("Cells can hold a maximum of 32,767 characters.");
```
Comment (Pankraty) [inline ClosedXML/Excel/Cells/XLCell.cs]: BTW, `ArgumentOutOfRangeException` would be more specific and it is inherited from `ArgumentException` so it is safe to replace it (no external code would be broken). Same for dates.
Up to you.

File: ClosedXML/Excel/Cells/XLCell.cs (hunk 2 above)
Comment (igitur) [reply ClosedXML/Excel/Cells/XLCell.cs]: done.

**Comment by igitur** [comment]
Ready for review.

File: ClosedXML/Excel/Cells/XLCell.cs [hunk 5]
```diff
@@ -2021,85 +2098,69 @@ private void ClearMerged()
...
             }
-            if (val.Length > 32767) throw new ArgumentException("Cells can only hold 32,767 characters.");
 
-            if (SetTableHeaderValue(val)) return;
-            if (SetTableTotalsRowLabel(val)) return;
+            if (SetTableHeaderValue(parsedValue)) return;
+            if (SetTableTotalsRowLabel(parsedValue)) return;
+
+            SetInternalCellValueString(parsedValue, validate: true, parseToCachedValue: false);
+            CachedValue = null;
```
Comment (Pankraty) [inline ClosedXML/Excel/Cells/XLCell.cs]: `CachedValue` won't become null if `SetTableHeaderValue(parsedValue)` or `SetTableTotalsRowLabel(parsedValue)` succeed. I am not sure but this seems wrong. Better move it to the top where `FormulaA1` and `_richText` are reset.

Comment (igitur) [reply ClosedXML/Excel/Cells/XLCell.cs]: `SetTableHeaderValue` sets `XLTableField.Name`, which calls `XLCell.SetValue<T>`(T value, setTableHeader: false)`, which will set `CachedValue` to null.

**Pull request merged**